### DISCUSSION TITLE
Custom policy partially runs, sees elements that aren't there

### DISCUSSION THREAD

**Discussion by l3112** [aquasecurity/trivy#4254] [Q&A] [answered]
I have my custom policy and my .tf file in the same directory, 

Code used:
`trivy conf --policy . --namespaces morganza .`

It says 'fails to load policy from [.]', and somehow continues to run it anyway.

## Debug
2023-05-09T12:11:28.183-0500	DEBUG	Severities: ["UNKNOWN" "LOW" "MEDIUM" "HIGH" "CRITICAL"]
2023-05-09T12:11:28.228-0500	DEBUG	cache dir:  /Users/morganza/Library/Caches/trivy
2023-05-09T12:11:28.228-0500	INFO	Misconfiguration scanning is enabled
2023-05-09T12:11:28.228-0500	DEBUG	Policies successfully loaded from disk
2023-05-09T12:11:28.272-0500	DEBUG	Walk the file tree rooted at '.' in parallel
2023-05-09T12:11:28.272-0500	DEBUG	Scanning Terraform files for misconfigurations...
2023-05-09T12:11:28.276-0500	FATAL	filesystem scan error:
    github.com/aquasecurity/trivy/pkg/commands/artifact.Run
        github.com/aquasecurity/trivy/pkg/commands/artifact/run.go:436
  - scan error:
    github.com/aquasecurity/trivy/pkg/commands/artifact.(*runner).scanArtifact
        github.com/aquasecurity/trivy/pkg/commands/artifact/run.go:267
  - scan failed:
    github.com/aquasecurity/trivy/pkg/commands/artifact.scan
        github.com/aquasecurity/trivy/pkg/commands/artifact/run.go:690
  - failed analysis:
    github.com/aquasecurity/trivy/pkg/scanner.Scanner.ScanArtifact
        github.com/aquasecurity/trivy/pkg/scanner/scan.go:146
  - post analysis error:
    github.com/aquasecurity/trivy/pkg/fanal/artifact/local.Artifact.Inspect
        github.com/aquasecurity/trivy/pkg/fanal/artifact/local/fs.go:166
  - post analysis error:
    github.com/aquasecurity/trivy/pkg/fanal/analyzer.AnalyzerGroup.PostAnalyze
        github.com/aquasecurity/trivy/pkg/fanal/analyzer/analyzer.go:487
  - terraform scan error:
    github.com/aquasecurity/trivy/pkg/fanal/analyzer/config.(*Analyzer).PostAnalyze
        github.com/aquasecurity/trivy/pkg/fanal/analyzer/config/config.go:47
  - scan config error:
    github.com/aquasecurity/trivy/pkg/misconf.(*Scanner).Scan
        github.com/aquasecurity/trivy/pkg/misconf/scanner.go:138
  - failed to load rego policies from [.]: 2 errors occurred**

**the two errors are `var cannot be used for rule name` in reference to the first line here:

```
allow bucket_name if {
  
	regex.match("ca_*", bucket_name)
	
}
```

- The file it's scanning doesn't have var either.
-  Nor do either package or policy
-  Is this scanning the built-in policies in conjunction to the custom one? But when I scanned them before making this custom policy, I didn't get the 'var' error either.

**Policy and config files:**
https://github.com/l3112/scan-trivy-issue/tree/main

**Will have slow/no responses from May 12th - May 15th.** 

(This community has been extremely helpful, thank you)

**Comment by simar7** [comment]
Could you share your custom policy and configuration files so we can try to investigate?

**Reply by l3112** [reply]
Of course:

https://github.com/l3112/scan-trivy-issue/tree/main

**Comment by simar7** [comment, marked as answer]
I assume you're trying to write a custom policy against terraform config. I've created a small repo here to get you started https://github.com/simar7/custom-policy-with-terraform

The errors you are receiving are rego errors and not Trivy errors. You can run your policies in the OPA playground to try them out: https://play.openpolicyagent.org

**Reply by l3112** [reply]
Thank you -- The only reason I'm unmarking it as answer is because I may have more questions / will be travelling before I can respond thoughtfully and I don't want to make too many topics if this ends up locked.
